### PR TITLE
make graphviz svg interactive

### PR DESCRIPTION
**Pull request by gertjanvanzwieten**

### CODE REVIEW

File: nutils/_graph.py [hunk 1]
```diff
@@ -68,7 +69,33 @@ def export_graphviz(self, *, fill_color: Optional[GraphvizColorCallback] = None,
...
+                graph = re.sub(b'<svg (.*?)>', insert_clickHandler, graph, count=1, flags=re.DOTALL)
+            img.write(graph)
+
+insert_clickHandler = rb'''<svg \1 onload="document.addEventListener('click', clickHandler)">
+<style>
+.highlight path { stroke: orange; stroke-width: 2; }
+.highlight polygon:first-of-type { fill: orange; }
+.highlight polygon { stroke: orange; }
+.highlight text { fill: white; }
+</style>
```
Comment (joostvanzwieten) [inline nutils/_graph.py:88]: Suggestion: With `g.edge, g.node { cursor: pointer }` you get a pointer if you hover above clickable elements.

File: nutils/_graph.py [hunk 2]
```diff
@@ -68,7 +69,33 @@ def export_graphviz(self, *, fill_color: Optional[GraphvizColorCallback] = None,
...
+<script>
+<![CDATA[
+function clickHandler(event) {
+    const g = event.target.closest("g");
```
Comment (joostvanzwieten) [inline nutils/_graph.py:92]: Suggestion: You can automatically unhighlight previously highlighted elements by

```javascript
for (const elem of Array.from(document.getElementsByClassName('highlight'))) {
    elem.classList.remove('highlight'); }
```

File: nutils/_graph.py [hunk 3]
```diff
@@ -68,7 +69,33 @@ def export_graphviz(self, *, fill_color: Optional[GraphvizColorCallback] = None,
...
+    if (g.classList.contains("edge")) {
+        g.classList.toggle("highlight"); }
+    else if (g.classList.contains("node")) {
+        const index = g.firstElementChild.textContent;
+        const pattern = new RegExp('(^|[^0-9])' + index + '($|[^0-9])')
```
Comment (joostvanzwieten) [inline nutils/_graph.py:97]: Nitpick: There's a semicolon missing.

**Review by joostvanzwieten — CHANGES_REQUESTED** [review]

**Comment by joostvanzwieten** [comment]
Very nice!

File: nutils/_graph.py (hunk 2 above)
Comment (gertjanvanzwieten) [reply nutils/_graph.py:92]: I often find myself tracing axes through the tree, highlighting the entire path, so I actually find it useful not to do this.

**Comment by gertjanvanzwieten** [comment]
> Very nice!

What about the scale modification? (2nd commit)

**Comment by joostvanzwieten** [comment]
> > Very nice!
> 
> What about the scale modification? (2nd commit)

Fine by me.

**Comment by gertjanvanzwieten** [comment]
> > > Very nice!
> > 
> > 
> > What about the scale modification? (2nd commit)
> 
> Fine by me.

Actually in the new version it's less ugly. Without onload I can insert the javascript near the bottom which is easier, and separate the scaling of width and height entirely (no longer confined to svg attributes, but it's probably ok)

File: nutils/_graph.py [hunk 4]
```diff
@@ -68,7 +69,33 @@ def export_graphviz(self, *, fill_color: Optional[GraphvizColorCallback] = None,
...
+        const isnew = g.classList.toggle("highlight");
+        for (const edge of document.getElementsByClassName('edge')) {
+            if (pattern.test(edge.firstElementChild.textContent)) {
+                edge.classList.toggle("highlight", isnew); } } } }
+]]>
+</script>'''
```
[the remark this answers is not in the document]
Comment (gertjanvanzwieten) [reply nutils/_graph.py:100]: done

File: nutils/_graph.py [hunk 5]
```diff
@@ -68,7 +69,33 @@ def export_graphviz(self, *, fill_color: Optional[GraphvizColorCallback] = None,
...
+
+insert_clickHandler = rb'''<svg \1 onload="document.addEventListener('click', clickHandler)">
+<style>
+.highlight path { stroke: orange; stroke-width: 2; }
+.highlight polygon:first-of-type { fill: orange; }
+.highlight polygon { stroke: orange; }
+.highlight text { fill: white; }
+</style>
+<script>
+<![CDATA[
```
[the remark this answers is not in the document]
Comment (gertjanvanzwieten) [reply nutils/_graph.py]: done

File: nutils/_graph.py [hunk 6]
```diff
@@ -68,7 +69,33 @@ def export_graphviz(self, *, fill_color: Optional[GraphvizColorCallback] = None,
                 for i, line in enumerate(src.split('\n'), 1):
                     print('{:04d}  {}'.format(i, line))
                 treelog.warning('graphviz failed for error code', status.returncode)
-            img.write(status.stdout)
+            graph = status.stdout
+            if image_type == 'svg':
+                graph = re.sub(b'<svg (.*?)>', insert_clickHandler, graph, count=1, flags=re.DOTALL)
+            img.write(graph)
+
+insert_clickHandler = rb'''<svg \1 onload="document.addEventListener('click', clickHandler)">
```
[the remark this answers is not in the document]
Comment (gertjanvanzwieten) [reply nutils/_graph.py]: done (super!)

File: nutils/_graph.py (hunk 3 above)
Comment (gertjanvanzwieten) [reply nutils/_graph.py]: fixed

File: nutils/_graph.py (hunk 1 above)
Comment (joostvanzwieten) [reply nutils/_graph.py:87]: Ah, there is a semicolon missing after pointer. Sorry!

File: nutils/_graph.py [hunk 7]
```diff
@@ -68,7 +68,33 @@ def export_graphviz(self, *, fill_color: Optional[GraphvizColorCallback] = None,
...
                     print('{:04d}  {}'.format(i, line))
                 treelog.warning('graphviz failed for error code', status.returncode)
-            img.write(status.stdout)
+            graph = status.stdout
+            if image_type == 'svg':
+                i = graph.rindex(b'</svg>')
+                graph = graph[:i] + clickHandler + graph[i:]
+            img.write(graph)
+
+clickHandler = rb'''
```
Comment (joostvanzwieten) [inline nutils/_graph.py:80]: This string doesn't have to be a raw string anymore.

File: nutils/_graph.py (hunk 1 above)
Comment (gertjanvanzwieten) [reply nutils/_graph.py:87]: fixed!

**Review by joostvanzwieten — APPROVED** [review]

**Pull request merged**